package.json

```diff
@@ -2712,6 +2796,37 @@
           }
         }
       }
-    ]
+    ],
+    "languages": [
+      {
+        "id": "aspnetcorerazor",
+        "extensions": [
+          ".cshtml"
```

Comment:
We're overriding VS Code's built-in `.cshtml` handling by mapping it to a different language ID. This separate `aspnetcorerazor` language ID lets us control exactly what behaviors are applied to it, so we can give a better experience (e.g., around HTML completion) that is more deeply aware of Razor language semantics.

Reply:
Shoot, just thought of one side effect of this. If the `omnisharp.preview` flag is false but we're still overriding the normal Razor language id I think this will nuke some of the default .cshtml editing experiences (no HTML completion).

Does VSCode allow us to toggle the language on/off?

Reply:
That's a great point. Unfortunately I don't have any easy fix in mind for it.

I checked, and as expected, having `omnisharp.preview` set to `false` means that `.cshtml` files:

* Still get correct colorization, since that's defined declaratively
* No longer get completions for HTML/JS
* No longer get automatic closing-tag insertion for HTML

As such it's still a working experience, but it is degraded in functionality compared with the default.

Possible ways forwards:

* Decide that Razor support is looking good already and turn it on by default in OmniSharp-VSCode
* Or, move Razor support out into a completely separate extension (not the installation experience we wanted)
* Or, ask the VS Code team to provide some feature by which an extension can respond to the `onLanguage:(languageId)` event and programmatically decline the event, causing VS Code to drop that `language` entry from its config and proceed to activate the next extension that matches the selector. Or any other design the VS Code team prefers.
    * Then we wait for it to be implemented and ship
* Or, try to implement our `.cshtml` support *without* defining a new language ID.
    * Unfortunately that is what we tried originally and it didn't work well because VS Code consulted both our extension and the built-in Razor feature for completions (etc), leading to duplication in the UI and incorrect awareness of what should show up in HTML blocks vs C# blocks etc. For example, if you try to type `List<string>` in a C# context, the built-in HTML tag completion kicks in and inserts `</string>` right after, which is insane.
    * At the very least, going this route is likely to involve some work with and by the VS Code folks to find some new way of suppressing the default behaviors programmatically.

Reply:
Assuming you don't think things are already good enough that you just want to turn it on, have you thought about maybe having a 'razor' branch in this repo where you can release beta versions of this extension with the razor support enabled, and folks who want to try it can install this new extension rather than toggling an option?

Reply:
@SteveSandersonMS lets talk more offline in our sync up today.

@gregg-miskelly we have. Our goals are to get the extension into as many peoples hands as possible. Forcing users to go and download a pre-release version would be an experience blocker and would prohibit us from getting enough feedback to make educated decisions about future releases.